### PR TITLE
Default main for ghcide 

### LINKED ISSUE #1336: extract defaultMain for ghcide
I'm working on this.

### PR DESCRIPTION
**Pull request by pepeiborra**
Fixes #1336

### CODE REVIEW

**Comment by wz1000** [comment]
I thought the goal was sharing the main function between HLS and ghcide?

File: ghcide/exe/Main.hs [hunk 1]
```diff
@@ -83,171 +52,46 @@ main = do
...
-                    { optShakeProfiling    = argsShakeProfiling
-                    -- , optOTMemoryProfiling = IdeOTMemoryProfiling argsOTMemoryProfiling
-                    , optTesting           = IdeTesting argsTesting
-                    , optShakeOptions      = (optShakeOptions defOptions){shakeThreads = argsThreads}
-                    , optCheckParents      = NeverCheck
-                    , optCheckProject      = False
+      _ -> runWithDb dbLoc $ \hiedb hiechan ->
+              Main.defaultMain (Main.defArguments hiedb hiechan) {
+                 Main.argFiles = case argFilesOrCmd of
+                    Typecheck x | argLSP -> Just x
```
Comment (wz1000) [inline ghcide/exe/Main.hs:74]:
```suggestion
                    Typecheck x | not argLSP -> Just x
```

**Comment by pepeiborra** [comment]
> I thought the goal was sharing the main function between HLS and ghcide?

That's the stretch goal. The two seem to have diverged, so I think it's better to go step by step and do it in separate PRs

**Review by wz1000 — APPROVED** [review]

**Comment by pepeiborra** [comment]
@wz1000 I have implemented the stretch goal too.

File: src/Ide/Main.hs [hunk 2]
```diff
@@ -107,114 +78,24 @@ hlsLogger = G.Logger $ \pri txt ->
...
-    pid <- T.pack . show <$> getProcessID
-    let
-        (plugins, commandIds) = ghcIdePlugins pid idePlugins
-        options = def { LSP.executeCommandCommands = Just commandIds
-                      , LSP.completionTriggerCharacters = Just "."
-                      }
-
-    if argLSP then do
-        t <- offsetTime
+    when argLSP $ do
```
Comment (wz1000) [inline src/Ide/Main.hs:88]: What happens if this is not set?

Comment (pepeiborra) [reply src/Ide/Main.hs:88]: See line 97 below

**Comment by pepeiborra** [comment]
A ghc 8.10.4 install failed due to a ghcup segfault:
```
  Attempting to install ghc 8.10.4 using ghcup
26
  /opt/hostedtoolcache/ghcup/0.1.12/x64/ghcup install ghc 8.10.4
27
  ghcup: internal error: Itimer: read(timerfd) failed with Bad file descriptor and returned -1
28
      (GHC version 8.8.4 for x86_64_alpine_linux)
29
      Please report this as a GHC bug:  https://www.haskell.org/ghc/reportabug
```

**Comment by konn** [comment]
Gresat! So, once this PR have landed, we can break down our monolithic func-test into per-plugin ones, right?

**Comment by pepeiborra** [comment]
@jneira the 8.10.4 func-test suite is failing 

> Gresat! So, once this PR have landed, we can break down our monolithic func-test into per-plugin ones, right?

Yeah, this PR will make it simpler to run ghcide with a custom set of HLS plugins for use in tests

**Comment by jneira** [comment]
mmm it worked in master and previous pr's, no?
I will take a look
do you want to test 8.10.3 again?

**Comment by pepeiborra** [comment]
> mmm it worked in master and previous pr's, no?
> I will take a look
> do you want to test 8.10.3 again?

Never mind, it was an issue with the PR. The problem now is that the MacOs builds are failing across all my PRs with a weird  error: 

image

image

**Pull request merged**